Sources/Utility/ArgumentParser.swift

```diff
@@ -382,12 +378,104 @@ public final class ArgumentParser {
+            case .oneByOne:
+                var parserProtocol = parser as ArgumentParserProtocol
+                try values.append(kind.init(parser: &parserProtocol))
+                parser = parserProtocol as! Parser
+            case .upToNextOption:
+                /// Iterate over arguments until the end or an optional argument
+                while true {
+                    let previousIterator = parser.argumentsIterator
+
+                    /// If there are no more arguments, we can stop parsing
+                    guard let nextArgument = try? parser.next() else {
+                        break
+                    }
+
+                    /// Parse the next argument only if it is non-optional
+                    if isPositional(argument: nextArgument) {
+                        try values.append(kind.parse(argument: nextArgument))
+                    } else {
+                        /// If we have reached an optional argument, we need to back-track by
+                        /// resetting the iterator to before the last `next` so that the
+                        /// top-level parser can continue from this argument.
+                        parser.argumentsIterator = previousIterator
+                        break
+                    }
+                }
+            case .remaining:
+                while let nextArgument = try? parser.next() {
+                    try values.append(kind.parse(argument: nextArgument))
+                }
+            }
+            
+            return values
+        }
+    }
+
+    /// A struct for the parser representation of option arguments.
+    struct ParserOptionArgument: ParserArgument {
+        let argument: AnyArgument
+        let strategy: ArrayParsingStrategy
+
+        func parse(with parser: inout Parser) throws -> [ArgumentKind] {
+            if let associatedArgument = parser.associatedArgumentValue {
+                return try [argument.kind.parse(argument: associatedArgument)]
+            }
+
+            let values = try strategy.parse(with: &parser, kind: argument.kind)
+
+            guard !values.isEmpty else {
+                throw ArgumentParserError.expectedValue(option: argument.name)
+            }
+
+            return values
+        }
+    }
+
+    /// A struct for the parser representation of positional arguments.
+    struct ParserPositionalArgument: ParserArgument {
+        let argument: AnyArgument
+        let strategy: ArrayParsingStrategy
+        let isOptional: Bool
+
+        func parse(with parser: inout Parser) throws -> [ArgumentKind] {
+            let value = try argument.kind.parse(argument: parser.currentArgument)
+            var values = [value]
+
+            switch strategy {
+            case .oneByOne:
+                break
+            case .upToNextOption,
+                 .remaining:
+                try values.append(contentsOf: strategy.parse(with: &parser, kind: argument.kind))
+            }
+
+            return values
```

Comment:
Again, these should be part of OptionArgument and PositionalArgument structures.